src/tcp_ip.cpp

```diff
@@ -0,0 +1,239 @@
+// Include custom TCP/IP API
+#include "tcp_ip.h"
+
+#if WIFI_ENABLED
+
+// Include FreeRTOS task API
+#include "freertos/task.h"
+// Include light-weight IP socket API
+#include "lwip/sockets.h"
+// Include custom Menu class implementation
+#include "menu.h"
+
+// ====================================== //
+// Define useful constants and data types //
+// ====================================== //
+
+// Define the number of currently supported TCP commands
+#define NUM_TCP_COMMANDS 3
+
+// Define, when receiving a TCP packet, what special strings should cause what actions
+typedef struct tcp_command_s {
+    // The string, that if the TCP packet matches, should trigger an action
+    const char *command;
+    // The action to be triggered if the TCP packet matched command
+    void (*action)();
+} tcp_command_t;
+
+// ======================= //
+// Instantiate useful data //
+// ======================= //
+
+// Intended to be read-only.
+// Keep track of the special strings that if a TCP packet matches,
+// should execute an action, and what action they should execute
+const tcp_command_t tcp_commands[NUM_TCP_COMMANDS] = {
+    {
+        .command = "up",
+        .action = []() { add_to_menu_input_queue(
+            /* MENU_INPUT_t menu_input = */ MENU_INPUT_UP,
+            /* bool from_isr = */ false); },
+    },
+    {
+        .command = "down",
+        .action = []() { add_to_menu_input_queue(
+            /* MENU_INPUT_t menu_input = */ MENU_INPUT_DOWN,
+            /* bool from_isr = */ false); },
+    },
+    {
+        .command = "confirm",
+        .action = []() { add_to_menu_input_queue(
+            /* MENU_INPUT_t menu_input = */ MENU_INPUT_CONFIRM,
+            /* bool from_isr = */ false); },
+    },
+#if 0
+    {
+        .command = "sleep",
+        .action = []() { add_to_menu_input_queue(
+            /* MENU_INPUT_t menu_input = */ MENU_INPUT_SLEEP,
+            /* bool from_isr = */ false); },
+    },
+#endif
+};
+
+// Keep track of the file handle being used as the IP socket connection
+// R | W | function
+// --+---+------------------
+//   |   |               
+//   |   |             
+// TODO: Put a mutex around this file handle
+int ip_socket_file_descriptor = 0;
+// Keep track of the handle of the task that reads IP packets (task_read_ip_packets(...))
+// R | W | function
+// --+---+------------------
+//   |   |               
+//   |   |             
+// TODO: Does this need a mutex?
+TaskHandle_t read_ip_packet_task_handle = nullptr;
```

Comment:
change to `NOTE: For now this code is good enough. Only *sleep*(...) and main(...) call it, if they never run at the same time, never in parallel. But, will need to update this code to be thread-safe if that changes.`

Reply:
addressed in: https://github.com/jCallon/squirt/pull/1/commits/7c45fb1afcd46a0976bd7b50a193a305785b00db